### ISSUE TITLE
Include specific machines leaks across tenants

### ISSUE THREAD

**Issue by michaelnoonan** [OctopusDeploy/Issues#3036]
# Steps to reproduce 

1. Create two tenants: `TenantA` and `TenantB`
2. Create two Cloud Regions, and allocate one target to `TenantA` and the other target to `TenantB`
3. Create a project, with one step to write out the Tenant name and Machine name.
4. Create a release and deploy it to both `TenantA` and `TenantB` using a basic deployment (let the machines be picked automatically by Octopus). It should work as expected, only targeting the correct machines.
5. Using the same release, deploy it to both `TenantA` and `TenantB` using an advanced deployment, but now `Include specific machines` to include both targets. Now the deployment will end up targeting both targets for both tenants. **This is a leak because we should never deploy TenantA's software onto a TenantB machine**

# Reproduction on demo.octopus.com

I've reproduced this issue on Octopus `3.7.11` here: https://demo.octopus.com/app#/projects/miketenantedproject

This deployment should have only included machines viable for Tenant A but targeted Tenant B's machines also: https://demo.octopus.com/app#/projects/Projects-761/releases/0.0.1/deployments/Deployments-244884

# Source

Source: http://help.octopusdeploy.com/discussions/problems/50093-steps-running-on-wrong-servers-tenants

**Comment by michaelnoonan** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Include specific machines leaks across tenants](https://trello.com/c/6ZDeF5Tk/24-include-specific-machines-leaks-across-tenants)

**Comment by michaelnoonan** [comment]
I've been trying to reproduce this issue since it was reported, and I think it has been fixed. I cannot reproduce it on Octopus 3.16, and haven't been able to reproduce it since 3.8.

I've created three Tenants:
- Germany (Tag: `Axis`)
- Great Britain (Tag: `Allies`)
- USA (Tag: `Allies`)

And an environment with:
- A dedicated target per Tenant
- A shared target for all Allies (by tag)
- A shared target for all Axis (by tag)

When I create multiple tenanted deployments in the one hit with the following "Include specific targets":

![image](https://user-images.githubusercontent.com/1627582/29200410-62a2ac08-7e98-11e7-9107-2c7028a5eece.png)

It works as expected (albeit with not the best error message):

![image](https://user-images.githubusercontent.com/1627582/29200421-744437b0-7e98-11e7-829f-0d67b268a2e2.png)

I'm going to close this issue as unreproducible.
